data/txHandler_test.go

```diff
@@ -1097,6 +1099,80 @@ loop:
 	require.Equal(t, 0, handler.txCanonicalCache.Len())
 }
 
+type dupMockNetwork struct {
+	mocks.MockNetwork
+	relay []map[network.Peer]struct{}
+}
+
+func (network *dupMockNetwork) RelayArray(ctx context.Context, tag []protocol.Tag, data [][]byte, wait bool, except map[network.Peer]struct{}) error {
+	network.relay = append(network.relay, except)
+	return nil
+}
+
+// TestTxHandlerProcessIncomingRelay ensures txHandler does not relay duplicates
+func TestTxHandlerProcessIncomingRelayDups(t *testing.T) {
+	partitiontest.PartitionTest(t)
+	t.Parallel()
+
+	const numUsers = 100
+	log := logging.TestingLog(t)
+	log.SetLevel(logging.Info)
+
+	// prepare the accounts
+	addresses, secrets, genesis := makeTestGenesisAccounts(t, numUsers)
+	genBal := bookkeeping.MakeGenesisBalances(genesis, sinkAddr, poolAddr)
+	ledgerName := fmt.Sprintf("%s-mem", t.Name())
+	const inMem = true
+	cfg := config.GetDefaultLocal()
+	cfg.Archival = true
+	cfg.EnableTxBacklogRateLimiting = false
+	cfg.TxIncomingFilteringFlags = 1 // txFilterRawMsg
+	ledger, err := LoadLedger(log, ledgerName, inMem, protocol.ConsensusCurrentVersion, genBal, genesisID, genesisHash, nil, cfg)
+	require.NoError(t, err)
+	defer ledger.Close()
+
+	net := dupMockNetwork{}
+
+	tp := pools.MakeTransactionPool(ledger.Ledger, cfg, logging.Base())
+	backlogPool := execpool.MakeBacklog(nil, 0, execpool.LowPriority, nil)
+	opts := TxHandlerOpts{
+		tp, backlogPool, ledger, &net, "", crypto.Digest{}, cfg,
+	}
+
+	handler, err := MakeTxHandler(opts)
+	require.NoError(t, err)
+	handler.Start()
+	defer handler.Stop()
+
+	defer handler.txVerificationPool.Shutdown()
+	defer close(handler.streamVerifierDropped)
+
+	_, blob := makeTxn(1, 2, addresses, secrets)
+
+	type mockPeer struct {
+		id int
+	}
+
+	for i := 1; i <= 3; i++ {
+		msg := network.IncomingMessage{Data: blob, Sender: mockPeer{i}}
+		handler.processIncomingTxn(msg)
+	}
+
+	// wait until txn propagates in handler
+	var timeout time.Duration
+	waitTime := 10 * time.Millisecond
+	for tp.PendingCount() < 1 {
+		time.Sleep(waitTime)
+		timeout += waitTime
+		if timeout >= time.Second {
+			require.Fail(t, "timed out waiting for txn to propagate")
+		}
+	}
```

Comment:
You can use "Eventually" to block on a condition for up to a maximum time, failing if it exceeds the time:

https://pkg.go.dev/github.com/stretchr/testify/require#Eventually

Reply:
done